tests/codec_test.py

```diff
@@ -0,0 +1,257 @@
+import os
+import base64
+from pathlib import Path
+from typing import Type, TypeVar, Any
+
+import capnp
+from google.protobuf.message import Message
+
+from xconn.client import connect_anonymous
+from xconn import codec
+from xconn.types import Event
+from tests.schemas.profile_pb2 import ProfileCreate, ProfileGet
+
+
+class String(str):
+    pass
+
+
+class Base64Codec(codec.Codec[String]):
+    def name(self) -> str:
+        return "base64"
+
+    def encode(self, obj: String) -> str:
+        return base64.b64encode(obj.encode("utf-8")).decode("utf-8")
+
+    def decode(self, data: str, out_type: Type[String]) -> String:
+        return out_type(base64.b64decode(data.encode("utf-8")).decode())
+
+
+def test_base64_codec():
+    encoder = Base64Codec()
+    encoded = encoder.encode(String("hello"))
+    assert isinstance(encoded, str)
+
+    decoded = encoder.decode(encoded, String)
+    assert isinstance(decoded, String)
+    assert decoded == "hello"
+
+
+class ProtobufCodec(codec.Codec[Message]):
+    def name(self) -> str:
+        return "protobuf"
+
+    def encode(self, obj: Message) -> bytes:
+        return obj.SerializeToString()
+
+    def decode(self, data: bytes, out_type: Type[Message]) -> Message:
+        msg = out_type()
+        msg.ParseFromString(data)
+
+        return msg
+
+
+def test_rpc_object_protobuf():
+    session = connect_anonymous("ws://localhost:8080/ws", "realm1")
+    session.set_payload_codec(ProtobufCodec())
+
+    def inv_handler(profile: ProfileCreate) -> ProfileGet:
+        return ProfileGet(
+            id="123",
+            username=profile.username,
+            email=profile.email,
+            age=profile.age,
+            created_at="2025-10-28T17:00:00Z",
+        )
+
+    session.register_object("io.xconn.profile.create", inv_handler)
+    create_msg = ProfileCreate(username="john", email="john@xconn.io", age=25)
+
+    result = session.call_object("io.xconn.profile.create", create_msg, ProfileGet)
+    assert isinstance(result, ProfileGet)
+    assert result.username == "john"
+    assert result.email == "john@xconn.io"
+    assert result.age == 25
+    assert result.id == "123"
+    assert result.created_at == "2025-10-28T17:00:00Z"
+
+    session.leave()
+
+
+def test_pubsub_object():
+    session = connect_anonymous("ws://localhost:8080/ws", "realm1")
+    session.set_payload_codec(Base64Codec())
+
+    def event_handler(event: Event):
+        assert event.args[0] == "hello"
+
+    session.subscribe_object("io.xconn.object", event_handler, String)
+
+    session.publish_object("io.xconn.object", String("hello"))
+
+    session.leave()
+
+
+def test_pubsub_protobuf():
+    session = connect_anonymous("ws://localhost:8080/ws", "realm1")
+    session.set_payload_codec(ProtobufCodec())
+
+    def event_handler(event: Event):
```

Comment:
this func needs to handle that directly (of course the unerlying logic needs fixing)
```suggestion
    def event_handler(create: ProfileCreate):
```